py-polars/polars/io/database/_executor.py

```diff
@@ -312,12 +315,33 @@ def _inject_type_overrides(
 
         return schema_overrides
 
+    def _is_alchemy_async(self) -> bool:
+        """Check if the cursor/connection/session object is async."""
+        from sqlalchemy.ext.asyncio import (
+            AsyncConnection,
+            AsyncSession,
+            async_sessionmaker,
+        )
+
+        return isinstance(
+            self.cursor, (AsyncConnection, AsyncSession, async_sessionmaker)
+        )
+
     def _normalise_cursor(self, conn: Any) -> Cursor:
         """Normalise a connection object such that we have the query executor."""
         if self.driver_name == "sqlalchemy":
-            conn_type = type(conn).__name__
-            if conn_type in ("Session", "async_sessionmaker"):
-                return conn
+            from sqlalchemy.engine import Engine
+            from sqlalchemy.ext.asyncio import (
+                AsyncEngine,
+                AsyncSession,
+                async_sessionmaker,
+            )
+            from sqlalchemy.orm import Session, sessionmaker
+
+            if isinstance(  # check for all flavours of session object
+                conn, (AsyncSession, Session, async_sessionmaker, sessionmaker)
+            ):
```

Comment:
I would make this a `_is_session` and then the other one can be `_is_session and _is_async`. But hey, this is already a lot better. Feel free to press the green button whenever 👍

Reply:
Done - will merge once latest CI run is green 👌